graph-builder/src/registry.rs

```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
+            })
+        })
+}
+
 /// Fetches a vector of all release metadata from the given repository, hosted on the given
 /// registry.
-pub fn fetch_releases(registry: &str, repo: &str) -> Result<Vec<Release>, Error> {
-    let mut metadata = Vec::new();
-    for tag in fetch_tags(registry, repo)? {
-        metadata.push(Release {
-            source: format!(
-                "{}/{}:{}",
-                registry
-                    .trim_left_matches("https://")
-                    .trim_left_matches("http://"),
-                repo,
-                tag
-            ),
-            metadata: fetch_metadata(registry, repo, &tag)?,
+pub fn fetch_releases(
+    registry: &str,
+    repo: &str,
+    credentials_path: &Option<PathBuf>,
+) -> Result<Vec<Release>, Error> {
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(|| -> Result<Box<dyn std::io::Read>, Error> {
+            match credentials_path {
+                Some(path) => match File::open(path) {
+                    Ok(file) => Ok(Box::new(file)),
+                    Err(e) => Err(e.into()),
+                },
+                None => Ok(Box::new(std::io::empty())),
+            }
+        }()?)
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let authenticated_client = tcore
+        .run(authenticate_client(
+            &mut client,
+            format!("repository:{}:pull", &repo),
+        ))
+        .map_err(|e| format_err!("{}", e))?;
+
+    let tags = match tcore.run(
+        authenticated_client
+            // According to https://docs.docker.com/registry/spec/api/#listing-image-tags
+            // the tags should be ordered lexically but they aren't
+            .get_tags(&repo, Some(20))
+            .map_err(|e| format_err!("{}", e))
+            .collect(),
+    ) {
+        Ok(mut tags) => {
+            if tags.len() == 0 {
+                bail!("{}/{} has no tags", registry, repo);
+            }
+            tags.sort();
+            Ok(tags)
+        }
+        Err(e) => Err(e),
+    }?;
+
+    let releases = futures::stream::iter_ok(tags.to_owned().into_iter())
+        .and_then(|tag| {
+            trace!("processing: {}:{}", &repo, &tag);
+            authenticated_client
+                .has_manifest(&repo, &tag, None)
+                .join(authenticated_client.get_manifest(&repo, &tag))
+                .map(|(manifest_kind, manifest)| {
+                    get_layers_for_manifest_with_kind(manifest_kind, manifest, tag)
+                })
+                .map_err(|e| format_err!("{}", e))
+        })
+        .and_then(|tag_layer_digests| match tag_layer_digests {
+            Ok((tag, layer_digests)) => {
+                if layer_digests.len() == 0 {
+                    bail!("layer_digests list empty")
```

Comment:
This doesn't seem like an error condition. We can handle this gracefully.

Reply:
Setting the context for the discussion: this is the case when there are no layers available for a discovered tag, which seems highly unexpected. Do you think it's better to just log the errors and drive the whole chain to completion and yield an empty result

Reply:
Yes. Logging the warning and then just returning an empty vector is fine.